### DISCUSSION TITLE
Forcing OSGi module install/start in NBP application

### DISCUSSION THREAD

**Discussion by sjaenick** [apache/netbeans#5140] [Q&A] [answered]
Hi,

in my NBP application (currently based on NBP 16), I use both NetBeans as well as OSGi modules; now, after some
refactoring work, I have an OSGi module for my API and another one with the corresponding implementation; multiple
NetBeans modules directly depend on the API, but only the (nbm-)application itself has a dependency against the implementation bundle.

This currently results in a situation where e.g. the BundleActivator of the API bundle is successfully invoked, but the
implementation bundle seems to be never loaded(?)/started at all. It seems that NBP doesn't consider OSGi bundles
unless they are listed as a direct/transitive dependency of any of the NetBeans modules.

As a workaround, I can add the implementation bundle as a dependency to any of the NetBeans modules and everything
works as intended. Is there a better way?

**Comment by iajn** [comment, marked as answer]
Your cluster / bundle might have autoload enabled. From https://netbeans.apache.org/tutorials/nbm-emf.html

_"Autoload" means that a module is turned on only when needed. In contrast to regular modules, which require some manual action, the autoload modules are opaque for users and are managed solely by the infrastructure. As soon as there is a module needing, via its dependencies, an autoload module, the infrastructure enables it._

I don't have an app to test it, but:

If your app is Ant based, I think you need to re-import the cluster (and delete the old config and update_tracking files). From there you can select how the bundle should be activated. Example in the tutorial: https://netbeans.apache.org/tutorials/nbm-emf.html

If you app is Maven based, I think you can change it with the NBM plugin. There is some info here: https://bits.netbeans.org/mavenutilities/nbm-maven-plugin/nbm-maven-plugin/index.html
